src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -1174,21 +1182,28 @@ internal ImmutableArray<LocalInfo<TypeSymbol>> GetLocalInfo(StandaloneSignatureH
         }
 
         /// <exception cref="UnsupportedSignatureContent">If the encoded parameter type is invalid.</exception>
-        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info)
+        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info, bool isReturn)
         {
+            var allowedRequiredModifiers = AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute;
+            if (isReturn)
+            {
+                // PROTOTYPE(init-only): can we make this more restrictive (ie. disallow aside from the return value of a setter)?
+                allowedRequiredModifiers |= AllowedRequiredModifierType.System_Runtime_CompilerServices_IsExternalInit;
+            }
+
             info.CustomModifiers = DecodeModifiersOrThrow(
                 ref signatureReader,
-                AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute,
+                allowedRequiredModifiers,
                 out SignatureTypeCode typeCode,
-                out bool inAttributeFound);
+                out AllowedRequiredModifierType inAttributeFound);
 
             if (typeCode == SignatureTypeCode.ByReference)
             {
                 info.IsByRef = true;
                 info.RefCustomModifiers = info.CustomModifiers;
                 info.CustomModifiers = DecodeModifiersOrThrow(ref signatureReader, AllowedRequiredModifierType.None, out typeCode, out _);
             }
-            else if (inAttributeFound)
+            else if ((inAttributeFound & AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute) != 0)
```

Comment:
>(inAttributeFound & AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute) != 0 [](start = 21, length = 96)

I think it would be more robust to base the check on what we allow, rather than on what we disallow. Consider using the following condition instead:
```(inAttributeFound & ~AllowedRequiredModifierType.System_Runtime_CompilerServices_IsExternalInit) != 0